### PR TITLE
chore: ml-sys onboarding exercises

### PR DESCRIPTION
**Pull request by rb-determined-ai**
Added 2-autodiff.md, the second of a series of onboarding exercises for ml-sys developers.

### CODE REVIEW

File: onboarding/mlsys/2-autodiff.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
+# Part 2: Autodifferentiation and Optimizers
+
+In "Part 1: Calculating Gradients and Averaging Them Together", we learned how
+gradient descent works mathematically.
+
+In this exercise, you'll learn how an actual deep learning library implements
+gradient descent.  You'll also learn about what optimizers are in a deep
+learning framework, and you'll implement one yourself, in both PyTorch and
+Tensorflow.  And lastly, you'll
```
Comment (tara-det-ai) [inline onboarding/mlsys/2-autodiff.md:9]: something got cut off > the last sentence trails off

File: onboarding/mlsys/2-autodiff.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
+# Part 2: Autodifferentiation and Optimizers
+
+In "Part 1: Calculating Gradients and Averaging Them Together", we learned how
+gradient descent works mathematically.
```
Comment (tara-det-ai) [inline onboarding/mlsys/2-autodiff.md:4]: i would stick to either "you" or "we" throughout

File: onboarding/mlsys/2-autodiff.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+loop that simulates momentum-based training with Optimizer State Partitioning
+across two workers.  Feel free to do it in PyTorch or TensorFlow.
+
+No need for actual distributed mechanics; just do it inside a single process.
+But do make sure your code demonstrates how the optimizer state can be sharded
+so each worker only needs to store half of the momentum state.
+
+## Summary
+
+Upon completion of this exercise, you have learned:
```
Comment (tara-det-ai) [inline onboarding/mlsys/2-autodiff.md:395]: try either one of these 

Upon completion of this exercise, you will have learned:

or

Congratulations for completing this exercise! In this exercise, you learned:

File: onboarding/mlsys/2-autodiff.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+- What optimizers do, and implemented a momentum optimizer yourself.
+- Familiarized yourself with low-level PyTorch and TensorFlow code.
+- Implemented one stage of ZeRO-DP memory optimization yourself.
```
Comment (tara-det-ai) [inline onboarding/mlsys/2-autodiff.md:399]: ## Summary

Congratulations! In this exercise, you accomplished:

- You learned what optimizers do, and implemented a momentum optimizer yourself.
- Familiarized yourself with low-level PyTorch and TensorFlow code.
- Implemented one stage of ZeRO-DP memory optimization yourself.

File: onboarding/mlsys/2-autodiff.md (hunk 2 above)
Comment (tara-det-ai) [reply onboarding/mlsys/2-autodiff.md:4]: In "Part 1: Calculating Gradients and Averaging Them Together", you learned how
gradient descent works mathematically.

**Review by tara-det-ai — APPROVED** [review]

File: onboarding/mlsys/2-autodiff.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+layers ($g'(x)$), results in the gradient including the current layer ($h'(x)$).
+
+If $Z$ were particularly easy to calculate, it would be a way to calculate the
+total gradient incrementally, by calculating the equivalent $Z$ term for each
+layer and multiplying them together.  But is $Z$ easy to calculate for a layer?
+
+The answer turns out to be "yes".  $Z$ is the result of feeding the output of
+the previous layer $g(x)$, which is easy to compute, into $f'$, the derivative
+of the operator $f$, which is often trivial to obtain (for example, if $f(n) =
+n^2$ then $f'(n) = 2n$).
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:56]: Personal preference, but I found this section a little confusing. Seems like it's asking the reader to like pack `f'(g)` into whatever `Z` is, forget about `Z`s origin, and then rediscover that it is just `f'(g)` again.

I think you may be going for like a recursive picture here, which is the more standard viewpoint.  There is some "upstream gradient" which has been computed already and passed to us, which here is like the derivative of the loss with respect to `f`: `dL/df`.  If we are interested in the derivative of the loss with respect to `g`, then you just need to know `f'(g)`, from which you get `dL/dg = dL/df * f'(g)`.  And then you probably need derivatives of tensors from lower layers still, for which you can now start with the upstream gradient `dL/dg` and so on and so forth.

File: onboarding/mlsys/2-autodiff.md [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+And lastly, what is the product of all the numbers in the final column?
+
+### A Complete Example, In PyTorch
+
+Now let's have PyTorch do the same thing for us.
+
+Start with some tensors `x` and `y`.
+
+```python
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:103]: I understand you're going for pedagogy here, but it would be very unusual to set `requires_grad = True` on `x` or `y`, or on any tensor which is not going to be updated by the optimizer, since this would just lead to unnecessary caching and computation.  Maybe just a comment to that effect is enough?

File: onboarding/mlsys/2-autodiff.md [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+### Q3
+
+If you rerun your forward pass (don't reinitialize `m`, but start with
+`mx = x * m`), then rerun `loss.backward()`, what effect does it have on your
+`m.grad`?
+
+Fun fact: multiple forward passes and `backward()` calls without zeroing
+gradients in between is exactly how we implement gradient aggregation in
+`PyTorchTrial` (only we don't reuse the same data for both forward passes).
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:172]: Maybe worth stating that this behavior is a happy side effect of the need for gradients to sum in this way in order to compute backpropagation correctly in the first place. The summation is needed in cases where a parameter is used in multiple operations.

Might be too detailed, though.

File: onboarding/mlsys/2-autodiff.md [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+  model parameters.  SGD will leave you in the local minimum forever.
+
+- It may be unduly affected by bad data, causing you to keep a lower learning
+  rate (and take longer to train) to compensate.
+
+There are many more sophisticated gradient updates in existence.  We're going
+to discuss a simple addition to the basic formula, which is to add "momentum"
+to our gradient calculation.  Momentum means that the gradient we apply at each
+step is a combination of the gradient we calculated in that step and the
+gradient we calculated in the previous step.  The formula becomes:
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:218]: More like "the gradient we apply at each step is a combination of the gradient we calculated in that step and a weighted sum of the all the gradients we computed in previous steps" (where more weight is give to more recent gradients).

File: onboarding/mlsys/2-autodiff.md [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+    )
+```
+
+## Optimizers and Large Models
+
+As you have seen, optimizer state is often at least as large as model state,
+even for fairly optimizers strategies like momentum.  This effect is compounded
+with other advanced features like mixed-precision training, where mixed
+precision, where even though model parameters require less memory each, the
+per-parameter optimizer state memory requirement actually grows.
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:375]: nit: in mixed precision, you actually double the model and keep one "master copy" in the original precision and the other in lower precision, so you don't even get memory savings in terms of the model.

(The master copy weights are canonically included in the optimize-state memory budget, confusingly.)

File: onboarding/mlsys/2-autodiff.md [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+
+As a result, training large models often means dealing with very large
+optimizer states.  A popular strategy is the [ZeRO Memory Optimization](
```
Comment (garrett361) [inline onboarding/mlsys/2-autodiff.md:378]: Maybe mention that DeepSpeed is the library which implements the ZeRO strategies.

**Comment by rb-determined-ai** [comment]
I'm apparently not going to get around to finishing this, but I'll leave the branch available if anybody wants to resurrect it at some point in the future.
